File: nlox/Lox/Scanner.cs
class Scanner
{
  private string source;
  private List<Token> tokens = new List<Token>();

  private int start = 0;
  private int current = 0;
  private int line = 1;

  private static Dictionary<string, TokenType> keywords = new Dictionary<string, TokenType>
  {
    { "and",    TokenType.And },
    { "class",  TokenType.Class },
    { "else",   TokenType.Else },
    { "false",  TokenType.False },
    { "for",    TokenType.For },
    { "fun",    TokenType.Fun },
    { "if",     TokenType.If },
    { "nil",    TokenType.Nil },
    { "or",     TokenType.Or },
    { "print",  TokenType.Print },
    { "return", TokenType.Return },
    { "super",  TokenType.Super },
    { "this",   TokenType.This },
    { "true",   TokenType.True },
    { "var",    TokenType.Var },
    { "while",  TokenType.While },
  };

  public Scanner(string source)
  {
    this.source = source;
  }

  public List<Token> ScanTokens()
  {
    while (!IsAtEnd())
    {
      start = current;
      ScanToken();
    }

    tokens.Add(new Token(TokenType.Eof, "", null, line));

    return tokens;
  }

  private void ScanToken()
  {
    var c = Advance();
    switch (c)
    {
      case '(':
        AddToken(TokenType.LeftParen);
        break;
      case ')':
        AddToken(TokenType.RightParen);
        break;
      case '{':
        AddToken(TokenType.LeftBrace);
        break;
      case '}':
        AddToken(TokenType.RightBrace);
        break;
      case ',':
        AddToken(TokenType.Comma);
        break;
      case '.':
        AddToken(TokenType.Dot);
        break;
      case '-':
        AddToken(TokenType.Minus);
        break;
      case '+':
        AddToken(TokenType.Plus);
        break;
      case ';':
        AddToken(TokenType.Semicolon);
        break;
      case '*':
        AddToken(TokenType.Star);
        break;
      case '!':
        AddToken(Match('=') ? TokenType.BangEqual : TokenType.Bang);
        break;
      case '=':
        AddToken(Match('=') ? TokenType.EqualEqual : TokenType.Equal);
        break;
      case '<':
        AddToken(Match('=') ? TokenType.LessEqual : TokenType.Less);
        break;
      case '>':
        AddToken(Match('=') ? TokenType.GreaterEqual : TokenType.Greater);
        break;
      case '/':
        if (Match('/'))
        {
          // A comment goes until the end of the line
          while (Peek() != '\n' && !IsAtEnd())
          {
            Advance();
          }
        }
        else
        {
          AddToken(TokenType.Slash);
        }
        break;
      case ' ':
      case '\r':
      case '\t':
        // Ignore whitespace.
        break;
      case '\n':
        line++;
        break;
      case '"':
        StringToken();
        break;
      case 'o':
        if (Match('r'))
        {
          AddToken(TokenType.Or);
        }
        break;
      default:
        if (IsDigit(c))
        {
          NumberToken();
        }
        else if (IsAlpha(c))
        {
          IdentifierToken();
        }
        else
        {
          Lox.Error(line, "Unexpected character.");
        }
        break;
    }
  }

  private bool IsAtEnd() => current >= source.Length;

  private char Advance()
  {
    return source[current++];
  }

  private bool Match(char c)
  {
    if (IsAtEnd())
    {
      return false;
    }
    if (source[current] != c)
    {
      return false;
    }

    current++;
    return true;
  }

  private char Peek()
  {
    if (IsAtEnd())
    {
      return '\0';
    }
    return source[current];
  }

  private char PeekNext()
  {
    if (current + 1 >= source.Length)
    {
      return '\0';
    }
    return source[current + 1];
  }

  private static bool IsDigit(char c) =>
    c >= '0' && c <= '9';

  private static bool IsAlpha(char c) =>
    (c >= 'a' && c <= 'z') ||
    (c >= 'A' && c <= 'Z') ||
    (c == '_');

  private static bool IsAlphaNumeric(char c) =>
    IsDigit(c) || IsAlpha(c);

  private void AddToken(TokenType type)
  {
    AddToken(type, null);
  }

  private void AddToken(TokenType type, Object? literal)
  {
    var text = source.Substring(start, current - start);
    tokens.Add(new Token(type, text, literal, line));
  }

  private void StringToken()
  {
    while (Peek() != '"' && !IsAtEnd())
    {
      if (Peek() == '\n')
      {
        line++;
      }
      Advance();
    }

    if (IsAtEnd())
    {
      Lox.Error(line, "Unterminated string.");
      return;
    }

    // The closing '"'.
    Advance();

    // Trim the surrounding quotes.
    var value = source.Substring(start + 1, current - start - 2);
    AddToken(TokenType.String, value);
  }

  private void NumberToken()
  {
    while (IsDigit(Peek()))
    {
      Advance();
    }

    // Look for a fractional part.
    if (Peek() == '.' && IsDigit(PeekNext()))
    {
      // Consume the '.'.
      Advance();

      while (IsDigit(Peek()))
      {
        Advance();
      }
    }

    var value = source.Substring(start, current - start);
    AddToken(TokenType.Number, Double.Parse(value));
  }

  private void IdentifierToken()
  {
    while (IsAlphaNumeric(Peek()))
    {
      Advance();
    }

    var text = source.Substring(start, current - start);
    var type = TokenType.Identifier;
    if (keywords.ContainsKey(text))
    {
      type = keywords[text];
    }
    AddToken(type);
  }
}

File: tslox/src/scanner.ts
import { error } from "./error";
import type { TokenType } from "./token";
import { Token } from "./token";

function isDigit(c: string) {
  return /[0-9]/.test(c);
}

function isAlpha(c: string) {
  return /[a-zA-Z_]/.test(c);
}

function isAlphanumeric(c: string) {
  return isDigit(c) || isAlpha(c);
}

const KEYWORDS = new Map<string, TokenType>([
  ["and", "And"],
  ["class", "Class"],
  ["else", "Else"],
  ["false", "False"],
  ["fun", "Fun"],
  ["for", "For"],
  ["if", "If"],
  ["nil", "Nil"],
  ["or", "Or"],
  ["print", "Print"],
  ["return", "Return"],
  ["super", "Super"],
  ["this", "This"],
  ["true", "True"],
  ["var", "Var"],
  ["while", "While"],
]);

export default class Scanner {
  private start = 0;
  private current = 0;
  private line = 1;
  private tokens: Token[] = [];

  constructor(private readonly source: string) { }

  scanTokens() {
    while (!this.isAtEnd()) {
      this.start = this.current;
      this.scanToken();
    }

    this.tokens.push(new Token("Eof", this.line, ""));
    return this.tokens;
  }

  private isAtEnd() {
    return this.current >= this.source.length;
  }

  private scanToken() {
    const c = this.advance();
    switch (c) {
      case "(": {
        this.addToken("LeftParen");
        break;
      }
      case ")": {
        this.addToken("RightParen");
        break;
      }
      case "{": {
        this.addToken("LeftBrace");
        break;
      }
      case "}": {
        this.addToken("RightBrace");
        break;
      }
      case ",": {
        this.addToken("Comma");
        break;
      }
      case ".": {
        this.addToken("Dot");
        break;
      }
      case "-": {
        this.addToken("Minus");
        break;
      }
      case "+": {
        this.addToken("Plus");
        break;
      }
      case ";": {
        this.addToken("Semicolon");
        break;
      }
      case "*": {
        this.addToken("Star");
        break;
      }
      case "!": {
        this.addToken(this.match("=") ? "BangEqual" : "Bang");
        break;
      }
      case "=": {
        this.addToken(this.match("=") ? "EqualEqual" : "Equal");
        break;
      }
      case "<": {
        this.addToken(this.match("=") ? "LessEqual" : "Less");
        break;
      }
      case ">": {
        this.addToken(this.match("=") ? "GreaterEqual" : "Greater");
        break;
      }
      case "/": {
        if (this.match("/")) {
          while (this.peek() != "\n" && !this.isAtEnd()) {
            this.advance();
          }
        } else {
          this.addToken("Slash");
        }
        break;
      }
      case " ":
      case "\r":
      case "\t":
        break;
      case "\n": {
        this.line++;
        break;
      }
      case '"': {
        this.stringToken();
        break;
      }
      case "o": {
        if (this.match("r")) {
          this.addToken("Or");
        }
        break;
      }
      default: {
        if (isDigit(c)) {
          this.numberToken();
        } else if (isAlpha(c)) {
          this.identifierToken();
        } else {
          error(this.line, "Unexpected character.");
        }
        break;
      }
    }
  }

  private advance() {
    return this.source[this.current++];
  }

  private match(c: string) {
    if (this.isAtEnd()) {
      return false;
    }
    if (this.source[this.current] != c) {
      return false;
    }

    this.current++;
    return true;
  }

  private peek() {
    if (this.isAtEnd()) {
      return "\0";
    }
    return this.source[this.current];
  }

  private peekNext() {
    if (this.current + 1 >= this.source.length) {
      return "\0";
    }
    return this.source[this.current + 1];
  }

  private addToken(type: TokenType, literal?: unknown) {
    var text = this.source.substring(this.start, this.current);
    this.tokens.push(new Token(type, this.line, text, literal));
  }

  private stringToken() {
    while (this.peek() !== '"' && !this.isAtEnd()) {
      if (this.peek() === '\n') {
        this.line++;
      }
      this.advance();
    }

    if (this.isAtEnd()) {
      error(this.line, "Unterminated string.");
    }

    // The closing '"'.
    this.advance();

    // Trim the surrounding quotes.
    const value = this.source.substring(this.start + 1, this.current - 1);
    this.addToken("String", value);
  }

  private numberToken() {
    while (isDigit(this.peek())) {
      this.advance();
    }

    // Look for fractional part.
    if (this.peek() === "." && isDigit(this.peekNext())) {
      // Consume the '.'.
      this.advance();

      while (isDigit(this.peek())) {
        this.advance();
      }
    }

    const value = this.source.substring(this.start, this.current);
    this.addToken("Number", Number.parseFloat(value));
  }

  private identifierToken() {
    while (isAlphanumeric(this.peek())) {
      this.advance();
    }

    const text = this.source.substring(this.start, this.current);
    let type: TokenType = "Identifier";
    if (KEYWORDS.has(text)) {
      type = KEYWORDS.get(text)!;
    }
    this.addToken(type);
  }
};
